torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -281,8 +281,11 @@ void ProcessGroupAgent::shutdown() {
       }
     }
   }
-  threadPool_.waitWorkComplete();
   listenerThread_.join();
+  // Note: calling threadPool_.waitWorkComplete() after listenerThread.join() so
+  // that we can finish any possible work enqueued into the thread pool, before
+  // python RPC handler is shutdown (see shutdown in rpc/api.py).
+  threadPool_.waitWorkComplete();
```

Comment:
This can launch more sends? In that case, do we need to abort those?

Reply:
Hmm, this is a good point, though in the current code it looks like we do this after aborting pending sends as well, but technically more sends can be launched in `threadPool.waitworkComplete()` again. Also, when this line is called we would have set `rpcRunning_` to false, and the send code checks this, so the send would immediately be stopped. I think this is why I'm not able to get this scenario to show up in the tests.

What about the following order?
1) join listener thread --> so this node stops accepting new requests, and just has to flush out its old ones
2) abort all existing pending sends --> basically end the current pending sends, do not send it over RPC
3) call `threadPool.waitWorkComplete()` ---> here, even if we have additional sends, they will immediately be cancelled, since we already check `rpcRunning_` flag when waiting for a send.

Reply:
Above looks good to me for non-graceful shutdown. For graceful shutdown, do we need to wait for cleanup for dist autograd context? Can the cleanup message still stay in the threadPool while we are doing this? Do we need sth similar to `_delete_call_user_rrefs()` for dist autograd?

Reply:
@mrshenli For graceful shutdown and cleaning up dist autograd context, I thought we are okay, since the following will happen:
1) Exit dist autograd context, triggering messages to be sent for cleanup
2) Graceful shutdown, eventually calls `sync()`, which waits for all messages to be cleanly processed across all nodes, meaning that we will process dist autograd cleanup while waiting in sync()
3) Shutdown, which in graceful case, should not have any existing work when calling `threadPool.waitWorkComplete` (I have a WIP diff to check this, by ensuring that we won't abort any pending sends in graceful shutdown)

Reply:
I see, this should be OK for now. One thing is that TensorPipeAgent is expecting us to deprecate sync/join APIs. In that case, dist_autograd probably need its own way to clear all messages. So that application messages will be cleared by wait_all_workers, RRef messages will be cleared by `_delete_all_user_rrefs`, and dist autograd can add its own internal message cleanup function in `shutdown()`. This might be sufficient to get rid of join.